planner/core/task.go

```diff
@@ -951,6 +972,46 @@ func buildIndexLookUpTask(ctx sessionctx.Context, t *copTask) *rootTask {
 	return newTask
 }
 
+func extractRows(p PhysicalPlan) float64 {
+	f := float64(0)
+	for _, c := range p.Children() {
+		if len(c.Children()) != 0 {
+			f += extractRows(c)
+		} else {
+			f += c.statsInfo().RowCount
+		}
+	}
+	return f
+}
+
+func calcPagingCost(ctx sessionctx.Context, t *copTask) float64 {
+	sessVars := ctx.GetSessionVars()
+	indexRows := t.indexPlan.statsInfo().RowCount
+	expectCnt := float64(t.expectCnt)
+	sourceRows := extractRows(t.indexPlan)
+	// with paging, the scanned rows is always less than or equal to source rows.
+	expectCnt = math.Min(expectCnt, sourceRows)
+	seekCnt := float64(1)
+	if expectCnt > pagingGrowingSum {
+		// if the expectCnt is larger than pagingGrowingSum, calculate the seekCnt for the excess.
+		seekCnt += float64(int(8 + (expectCnt-pagingGrowingSum)/maxPagingSize))
+	} else if expectCnt > minPagingSize {
+		// if the expectCnt is less than pagingGrowingSum,
+		// calculate the seekCnt(number of terms) from the sum of a geometric progression.
+		// expectCnt = minPagingSize * (pagingSizeGrow ^ seekCnt - 1) / (pagingSizeGrow - 1)
+		// simplify (pagingSizeGrow ^ seekCnt - 1) to pagingSizeGrow ^ seekCnt, we can infer that
+		// seekCnt = log((pagingSizeGrow - 1) * expectCnt / minPagingSize) / log(pagingSizeGrow)
+		seekCnt += float64(int(math.Log((pagingSizeGrow-1)*expectCnt/minPagingSize) / math.Log(pagingSizeGrow)))
+	}
+	indexSelectivity := float64(1)
+	if sourceRows > indexRows {
+		indexSelectivity = indexRows / sourceRows
+	}
+	pagingCst := (seekCnt-1)*sessVars.GetSeekFactor(nil) + expectCnt*sessVars.CPUFactor
```

Comment:
Why the default seekCnt is 1 and use `seekCnt-1` here?

Reply:
Because what we want is the diff between non-paging cost and paging cost, without paging, there is `1` seek cost(The seek cost is not included in `idxCst`), and with paging, there is `seekCnt` seek cost, so there should be `seekCnt-1`.
The reason default `seekCnt` is there should be 1 seek at least.

Reply:
Does `expectCnt*sessVars.CPUFactor` include 1 seek cost for both paging or non-paging requests? But I still can't get the point to use 1 as the default value of `seekCnt` and use `seekCnt += ...` not `seekCnt = ...`.